drivers: adc: stm32: support multiple channels

heinwessels opened the pull request:
Support sequencing multiple channels into a single read.
Split off from #52965

Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

heinwessels committed:
drivers: adc: stm32: support multiple channels

Support sequencing multiple channels into a single read

Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

benjaminbjornsson commented on samples/drivers/adc/src/main.c:
```diff
@@ -62,18 +75,16 @@ void main(void)
 
-			(void)adc_sequence_init_dt(&adc_channels[i], &sequence);
-
 			err = adc_read(adc_channels[i].dev, &sequence);
```
Shouldn't this be moved outside the for loop? This would restrict `io-channels` to only contain channels from one device, but that might be ok?

heinwessels replied:
Same comment.

benjaminbjornsson replied:
Well, this commen ended up on the wrong line, I'll move it :)

benjaminbjornsson reviewed (commented):
Thank you for this work!

I only looked at the sample for now, the driver part is way above my head atm :)

heinwessels commented on samples/drivers/adc/src/main.c:
```diff
@@ -51,8 +57,15 @@ void main(void)
 			return;
 		}
+
+		(void)adc_sequence_init_dt(&adc_channels[i], &sequence);
```
You are correct, it is indeed odd. I'll move it outside the loop. 

It's still not ideal, although I think it's good enough for now. In the `dts` a configuration (resolution, oversampling, etc) is specified for each channel, but the `sequence` will only remember it for one, and the actual driver cannot change the settings for each channel. (Especially when DMA is enabled in a follow-up PR).

@erwango do you agree?

heinwessels replied:
Closing this conversation, as it's implemenation dependent. The sample application shows the current limitations of the driver, and improving the mechanism will require quite a bit of rework.

xyzzy42 commented on drivers/adc/adc_stm32.c:
```diff
@@ -758,71 +778,81 @@ static int start_read(const struct device *dev,
+			LL_ADC_SetChannelPreSelection(adc, channel);
 #elif defined(CONFIG_SOC_SERIES_STM32U5X)
-	/*
-	 * Each channel in the sequence must be previously enabled in PCSEL.
-	 * This register controls the analog switch integrated in the IO level.
-	 * Only for ADC1 instance (ADC4 has no Channel preselection capability).
-	 */
-	if (adc == ADC1) {
-		LL_ADC_SetChannelPreselection(adc, channel);
-	}
+			/*
+			 * Each channel in the sequence must be previously enabled in PCSEL.
+			 * This register controls the analog switch integrated in the IO level.
+			 * Only for ADC1 instance (ADC4 has no Channel preselection capability).
+			 */
+			if (adc == ADC1) {
+				LL_ADC_SetChannelPreselection(adc, channel);
+			}
 #endif
```
The ifdeffery here can be decreased by using `IS_ENABLED()` (which I thought was to be preferred anyway).  E.g., 
```c
if (IS_ENABLED(CONFIG_SOC_SERIES_STM32H7X) ||
    (IS_ENABLED(CONFIG_SOC_SERIES_STM32U5X) && adc == ADC1)) {
        LL_ADC_SetChannelPreselection(adc, channel);
}
```
Also less copypasta.

erwango replied:
This will work only if `LL_ADC_SetChannelPreselection` is defined on all other series.
Code under a `IS_ENABLED(0)` is compiled first and then linked out. If header is not known, it won't compile.

erwango replied:
Though, I totally agree there are some readability improvements that should be made on this driver.

xyzzy42 replied:
The HAL only provides the functions for specific series?  That's annoying, it defeats the modern style of not using ifdef as much.

A workaround would be to add an empty macro to define `LL_ADC_SetChannelPreselection()`.  It would be nicer if the HAL did that.

erwango replied:
I agree we need to heavily review this driver to make it more readable. But let's not make current PR more complex

xyzzy42 replied:
Yes, I agree out of scope.  It's not just a case of writing code being added in the PR differently.

heinwessels replied:
Sounds good to me too, it needs to be more readable but it's out of scope. Implementing @xyzzy42 suggestion for more efficentl looping will reduce the indent by one tab, which will make it more readable as well.

As a side note, IMO this driver needs an `adc_stm32.h` where the lookup tables and helper functions can move to. This will cleanup this driver quite a bit.

xyzzy42 commented on drivers/adc/adc_stm32.c:
```diff
@@ -758,71 +778,81 @@ static int start_read(const struct device *dev,
+	data->samples_count = 0;
 
-	adc_stm32_setup_channels(dev, index);
+	for (uint8_t channel_id = 0; channel_id < STM32_CHANNEL_COUNT; channel_id++) {
+		if (data->channels & BIT(channel_id)) {
```
Instead of looping for every bit, you can efficiently only loop on the set bits:
```c
uint8_t channel_id;
for (uint32_t channels = data->channels; channels; channels &= ~BIT(channel_id)) {
        channel_id = find_msb_set(channels) - 1;
}
```

heinwessels replied:
Added here and in `adc_stm32_teardown_channels` where it removed an extra indentation. It's a nice trick!

xyzzy42 commented on drivers/adc/adc_stm32.c:
```diff
@@ -984,8 +1012,11 @@ static void adc_stm32_isr(const struct device *dev)
+	data->samples_count++;
+	if (data->samples_count == data->channel_count) {
+		adc_context_on_sampling_done(&data->ctx, dev);
+		adc_stm32_teardown_channels(dev);
```
This seems unexpected to me.

If `adc_context_on_sampling_done()` calls the user call back and it returns `ADC_ACTION_REPEAT`, or the buffer is not finished because it has room for multiple samples, then `adc_context_on_sampling_done()` will cause `adc_context_start_sampling()` to get called again for another sample.

It seems odd that the channels get torn down here.  When `adc_context_start_sampling()` is called it will call `adc_stm32_start_conversion()` and that appears to re-enable the ADC.  It seems like the channels never get setup again.

heinwessels replied:
This seems to be an issue with the existing driver as well, where the channels are torn down but not configured again when `ADC_ACTION_REPEAT`. The existing driver might even teardown the channel before the read is done if in `async`.

The only way to teardown the channels at the correct time is to do it when `adc_context_complete`, but currently `adc_stm32.c` does not know when this occurs. We could add a `on_context_complete` callback that the stm32 driver can subscribe to?

xyzzy42 replied:
I think adc_context makes the assumption that if sampling is driven by a kernel timer, then it can just stop trigger when sampling is done.  There is no need to unconfigure the input mux or anything else.  So there is no hook into `adc_context_complete()`.

If the sampling is done by an ADC hardware timer (maybe a "continuous-mode" bit in the ADC), then it will need to be turned off when sampling is done.  But in this case the driver does not define `ADC_CONTEXT_USES_KERNEL_TIMER` and the adc_context code will give the driver a hook `adc_context_disable_timer()` where it can place code to run when the entire sequence is finished.

I can think of two ways to solve it:

1. I don't think the driver must teardown the channels.  It's not going to generate more interrupts.  So don't.  Move that feature to power management support.
2. It's reasonable that some ADCs need to do stuff other than turn off a timer when they are done and there should be hook in adc_context for this.  The style here would probably be a driver static function with a fixed `adc_context_*` name, rather than a function pointer, like how `adc_context_disable_timer()` is implemented.

heinwessels replied:
I would prefer leaving the teardown channels in, because it is useful low-power applications, as mentioned when it was added in https://github.com/zephyrproject-rtos/zephyr/commit/411079cf1d885747fcb9f84b27e168c09e895226.

I think we should go with the second option you list there, which is similar to my proposal except we're not storing a function pointer.

erwango replied:
> I think we should go with the second option you list there, which is similar to my proposal except we're not storing a function pointer.

@heinwessels If you do please think about doing it in a separate commit ;-)

xyzzy42 replied:
Slight variation on 2, have a macro, like the existing `ADC_CONTEXT_USES_KERNEL_TIMER`, that causes adc_context.h to not define `adc_context_complete()`.  The driver can do it.  It would have to duplicate the existing actions of that function, but it's not that much, and this way the driver would be able to decide if it should notify the thread waiting for the ADC data before or after it does the channel teardown, or other complications with how the sequence should finish.  So one exchanges code duplication for flexibility.

But whatever you decide, if you don't fix the driver in this PR, it would be nice to document that repeated sampling doesn't work, and check if `sequence->options->extra_samplings` is non-zero and return EINVAL.  No way to detect ADC_ACTION_REPEAT though.

heinwessels replied:
I've added a `__weak` function called `adc_context_on_complete` that can optionally be implemented by concrete drivers that is called from `adc_context_complete`. The `adc_stm32_teardown_channels` is then only called from that location.

heinwessels replied:
Although now I see a CI warning
```
 WEAK_DECLARATION: Using weak declarations can have unintended link defects
```
I'll do something similar to `ADC_CONTEXT_USES_KERNEL_TIMER` then.

heinwessels replied:
Okay, now it's a function enabled if `ADC_CONTEXT_ENABLE_ON_COMPLETE` is defined.

xyzzy42 commented on drivers/adc/adc_stm32.c:
```diff
@@ -626,69 +629,86 @@ static void adc_stm32_unset_common_path(const struct device *dev, uint32_t PathI
+				if ((__LL_ADC_CHANNEL_TO_DECIMAL_NB(
+					     LL_ADC_CHANNEL_TEMPSENSOR_ADC1) == channel_id) &&
+				    (config->base == ADC1)) {
+					adc_stm32_disable(adc);
```
This (and all the other copy-pasta of this same code block) will call `adc_stm32_disable(adc);` over and over in the loop.  Why not move the disable call outside the loop?

IMHO, a function that does a (channel_id, adc_id) →LL_ADC_PATH_constant mapping would make this code and the copy of it in setup_channels massively smaller and easier to read.  The ifdefs are out of control.

erwango replied:
Ageed. If you do, please use a specific commit.

heinwessels replied:
I moved the `adc_stm32_disable` to before the loop, and decrease the indentation by one. IMO this is is much more readable, and could maybe be cleaned up if required later when this driver is made more readable.

heinwessels replied:
I've decided that this is out of scope for this PR, and can be cleanup up in the inevitable refactor/cleanup of this driver. For now I've moved the `disable` back into the `if`s so that it's not called unnecessarily.

heinwessels committed:
drivers: adc: adc_context: add optional context_on_complete

This optional function allows a driver to do operations

after a complete sampling sequence is completed.

Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

heinwessels committed:
drivers: adc: stm32: only teardown channel after context complete

Previously channels could be tore down before a sequence completes,

either when using async, or when ADC_ACTION_REPEAT was specified

The teardown is now moved to after an entire context is complete

Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

heinwessels commented:
@erwango @xyzzy42 @FRASTM ready for re-review. 

Sorry if a double ping, I can't seem to trigger it for all three at the same time. Also lost some other review requests again. Github is being weird with the Reviewers pane.

heinwessels commented:
I also corrected `samples/drivers/adc/src/main.c` to correctly read and print a sequence of channels. Previously if a sequence was supplied it would read it all correctly, but repeat the entire sequence-read redundently for each channel.

FRASTM commented on drivers/adc/adc_stm32.c:
```diff
@@ -948,11 +969,9 @@ static int start_read(const struct device *dev,
 
-	err = adc_context_wait_for_completion(&data->ctx);
-
-	adc_stm32_teardown_channels(dev, index);
```
no more adc_stm32_teardown_channels()  ?

heinwessels replied:
Not removed, it moved to `adc_context_on_complete`.

heinwessels commented on drivers/adc/adc_stm32.c:
```diff
@@ -758,71 +772,78 @@ static int start_read(const struct device *dev,
-		return -ENOTSUP;
-	}
+	for (uint32_t channels = data->channels; channels; channels &= ~BIT(channel_id)) {
+		channel_id = find_lsb_set(channels) - 1;
```
That part of the for-loop-definition is only executed _after_ the first iterarion is completed, meaning any garbage initial value would be overwritten already. I had the same thought at first but tested and verified that it doesn't matter.

erwango commented:
@heinwessels We've launched a test session on our bench and `tests/drivers/adc/adc_api/drivers.adc` is failed on all boards. Can you have a check ?

xyzzy42 commented on drivers/adc/adc_stm32.c:
```diff
@@ -758,71 +772,78 @@ static int start_read(const struct device *dev,
 #else
-	LL_ADC_REG_SetSequencerRanks(adc, table_rank[0], channel);
-	LL_ADC_REG_SetSequencerLength(adc, table_seq_len[0]);
+		LL_ADC_REG_SetSequencerRanks(adc, table_rank[data->channel_count], channel);
```
`channel_count` is almost correct.  The ordinal position of the channel in the sequence should be used here.  `data->channels` is a bit mask, so the first channel in the sequence is not necessarily channel 0, and so on.

But I think it's off by 1.  The first channel should use table_rank[0] to be programmed as the first sequencer slot.  But data->channel_count is incremented at the start of loop, so it will be 1.

OTOH, it looks like `table_seq_len[0]` means zero channels in the sequence, so the correct index in the table is being used for that table.

heinwessels replied:
Indeed, it was off by one. I missed this change because the `data->channels++` was originally done at the end of the loop. And this applies to both `rank` and `seq_len`, because with a length of 1 the sequencer should be disabled. 

I changed how it's calculated to be easier to read by using a `channel_index` which is incremented at the end of the loop.

``` C
		LL_ADC_REG_SetSequencerRanks(adc, table_rank[channel_index], channel);
		LL_ADC_REG_SetSequencerLength(adc, table_seq_len[channel_index]);
```

xyzzy42 replied:
Without LTO, using local variables as loop indexes is better anyway.  The compiler has a hard time determining that no function calls will access `data->channels` when it can only see this one file.  So it must update the value in RAM.  But the compiler can see that no external functions can access the local variable so it can just leave it in a register.

xyzzy42 commented on drivers/adc/adc_stm32.c:
```diff
@@ -758,71 +772,78 @@ static int start_read(const struct device *dev,
+	}
+
+	if (data->channel_count == 0) {
+		LOG_ERR("No channels selected");
```
Quick check of STM32L5x show that sequencer has a limit of 16 channels.  I think one could read off end of the tables.

The channel bitmaks could be validated before the loop.

```
if (data->channels > BIT(STM32_CHANNEL_COUNT) - 1)
  // bitmask uses out of range channel
if (popcount(data->channels) > 16)
  // too many channels for sequencer
```
I'm sure the max number of channels in a sequence is device specific in a complex way that is annoying to deal with.

heinwessels replied:
Added these two checks.

The max number of channels I set to `ARRAY_SIZE(table_seq_len)`. That array already signifies the maximum, as all boards that has that included supports 16 sequencer length. It's only the `wlxx` and `g0xx` has max 8 channels in sequences, but they are `if !defined(...)` out.

heinwessels committed:
tests: drivers: adc: adc_api: add second channel to nucleo_h743zi

Adds a second channel to the nucleo_h743zi test
because it's now supported by the driver

Signed-off-by: Hein Wessels <heinwessels93@gmail.com>

heinwessels commented:
> @heinwessels We've launched a test session on our bench and `tests/drivers/adc/adc_api/drivers.adc` is failed on all boards. Can you have a check ?

@erwango Ah, sorry about that. This was caused by a problem with the sequencer/ranks as specified in a comment, and incorrect handling of `extra_samples`. I've fixed both issues, and added a `ADC_2ND_CHANNEL_ID` for the nucleo_h743zi to correctly do the unit tests (in a separate commit ;) ). The unit test shows all passing.

xyzzy42 commented on drivers/adc/adc_stm32.c:
```diff
@@ -758,71 +773,100 @@ static int start_read(const struct device *dev,
-	uint8_t index = find_lsb_set(channels) - 1;
+	data->buffer = sequence->buffer;
+	data->channels = sequence->channels;
+	data->channel_count = POPCOUNT(data->channels);
```
`POPCOUNT` and not `popcount`?  I only see the latter in the Zephyr codebase.

xyzzy42 commented on drivers/adc/adc_stm32.c:
```diff
@@ -984,12 +1026,25 @@ static void adc_stm32_isr(const struct device *dev)
-	adc_context_on_sampling_done(&data->ctx, dev);
+	data->samples_count++;
+	/* ISR is triggered after each conversion, and at the end-of-sequence. */
+	if (data->samples_count % data->channel_count == 0) {
```
Since `samples_count` is just 8 bits, the counter will possibly overflow.  This doesn't matter if `channel_count` is a power of two, but it doesn't need to be, and if it's not a power of two the result won't be correct.

I suggest:
```
if (data->samples_count == data->channel_count) {
        data->samples_count = 0;
}
```
Firstly, it easily avoids the overflow problem.  Secondly, the modulus (`data->channel_count`) is non-constant.  The compiler can come up with efficient code to find a constant modulus.  Obviously for a power of two it's just an AND operation, but even for numbers like 5 there is fast code.  But for a non-constant it has to do full integer division, which is very slow.  Usually, there is no integer division instruction (ARM), but even when there is (x86), it's by far the slowest ALU instruction in the CPU.  Even slower than floating point division surprisingly.

heinwessels replied:
Changed. Good catch :)

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -758,71 +772,100 @@ static int start_read(const struct device *dev,
+	uint8_t channel_id;
+	uint8_t channel_index = 0;
+
+	for (uint32_t channels = data->channels; channels;
+		      channels &= ~BIT(channel_id), channel_index++) {
```
Let's put a comment here, as this is part not obvious.

heinwessels replied:
Something like this?
``` C
	/* Iterate over selected channels in bitmask keeping track of:
	 * - channel_index: ranging from 0 -> ( data->channel_count - 1 )
	 * - channel_id: ordinal position of channel in data->channels bitmask
	 */
	for (uint32_t channels = data->channels; channels;
```

fabiobaltieri commented:
Is the sample still going to work on devices that do not support multiple channels in one read?

heinwessels commented:
> Is the sample still going to work on devices that do not support multiple channels in one read?

@fabiobaltieri  Yes, it should. The driver acts exactly the same as before when only one channel is selected. So to the LL drivers and the user there will be no difference.

And it's only possible to selected multiple channels if the board has an sequencer, which I used a previously existing check's logic (where it creates `table_seq_len`).

fabiobaltieri commented:
> @fabiobaltieri Yes, it should. The driver acts exactly the same as before when only one channel is selected. So to the LL drivers and the user there will be no difference.
> 
> And it's only possible to selected multiple channels if the board has an sequencer, which I used a previously existing check's logic (where it creates `table_seq_len`).

Right but isn't the sample always trying to select multiple channels if multiple channels are defined now? Thinking of a board with no sequencer and multiple channel defined, before they were going to be used one at a time, now it would try to do all at once. (Sorry not sure it's a problem and admittedly not very familiar with the API)

heinwessels commented:
> > @fabiobaltieri Yes, it should. The driver acts exactly the same as before when only one channel is selected. So to the LL drivers and the user there will be no difference.
> > And it's only possible to selected multiple channels if the board has an sequencer, which I used a previously existing check's logic (where it creates `table_seq_len`).
> 
> Right but isn't the sample always trying to select multiple channels if multiple channels are defined now? Thinking of a board with no sequencer and multiple channel defined, before they were going to be used one at a time, now it would try to do all at once. (Sorry not sure it's a problem and admittedly not very familiar with the API)

No problem, it's good to make sure. There's this piece of code that will return an error if more than one channel is specified in `sequence->channels` and the board doesn't support a sequencer. If it has only one channel defined, then it will configure that channel in the same way as before this PR. Here is a snippet of the relevant code:
```C
#if !defined(CONFIG_SOC_SERIES_STM32F0X) && \
	!defined(CONFIG_SOC_SERIES_STM32G0X) && \
	!defined(CONFIG_SOC_SERIES_STM32L0X) && \
	!defined(CONFIG_SOC_SERIES_STM32WLX)
	if (data->channel_count > ARRAY_SIZE(table_seq_len)) {
		LOG_ERR("Too many channels for sequencer. Max: %d", ARRAY_SIZE(table_seq_len));
		return -EINVAL;
	}
#else
	if (data->channel_count > 1) {
		LOG_ERR("This device only supports single channel sampling");
		return -EINVAL;
	}
#endif
```
So basically:
- Sampling a _single_ channel will setup the low-level drivers in the _exact same way_ as before, regardless of if board supports multiple channels or not.
- If a board doesn't support sampling multiple channels, it will not allow the user to specify multiple channels.

I think that's secure enough :)

anangl commented:
> > Is the sample still going to work on devices that do not support multiple channels in one read?
> 
> @fabiobaltieri Yes, it should. The driver acts exactly the same as before when only one channel is selected. So to the LL drivers and the user there will be no difference.
> 
> And it's only possible to selected multiple channels if the board has an sequencer, which I used a previously existing check's logic (where it creates `table_seq_len`).

No, the sample won't work in this form for example for ESP32 boards, where two channels are selected, like here: https://github.com/zephyrproject-rtos/zephyr/blob/7941a88121083c43eb500ca2a9a4dd82838b07dc/samples/drivers/adc/boards/esp32c3_devkitm.overlay#L11-L13 but the driver does not support multichannel sequences: https://github.com/zephyrproject-rtos/zephyr/blob/7941a88121083c43eb500ca2a9a4dd82838b07dc/drivers/adc/adc_esp32.c#L150-L153

fabiobaltieri commented:
>     * If a board doesn't support sampling multiple channels, it will not allow the user to specify multiple channels.

Right so a sample could potentially break? Like in the case that @anangl pointed out? (I have ~that hardware, can give it a try) If that's the case then I think we should either have two tests or two modes for the test.

heinwessels commented:
> No, the sample won't work in this form for example for ESP32 boards, where two channels are selected, like here:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/7941a88121083c43eb500ca2a9a4dd82838b07dc/samples/drivers/adc/boards/esp32c3_devkitm.overlay#L11-L13
> 
> but the driver does not support multichannel sequences:
> https://github.com/zephyrproject-rtos/zephyr/blob/7941a88121083c43eb500ca2a9a4dd82838b07dc/drivers/adc/adc_esp32.c#L150-L153

> Right so a sample could potentially break? Like in the case that @anangl pointed out? (I have ~that hardware, can give it a try) If that's the case then I think we should either have two tests or two modes for the test.

That's for a different family though? This PR only address multiple channel sampling to the `STM32` family, not `ESP32`. And all ADC drivers are built using `adc_context` which inherently supports multi-channel-sampling. It's up to the specific driver to implement it or protect against its accidental use if not.

So, honestly I'm a little lost with where this conversation went :)

fabiobaltieri commented:
> That's for a different family though? This PR only address multiple channel sampling to the `STM32` family, not `ESP32`. And all ADC drivers are built using `adc_context` which inherently supports multi-channel-sampling. It's up to the specific driver to implement it or protect against its accidental use if not.
> 
> So, honestly I'm a little lost with where this conversation went :)

Yeah but the sample is generic, it lives in `samples/drivers/adc`. Not to say it has to support every option, but it makes sense to have samples for bot modes. If it was stm32 specific it could go under `sapmle/boards` but in this case it's just a subsystem functionality not supported by all drivers, so I'd say that either the sample supports both configs you could copy it to a new sample and have only few controlled platforms it can run off. (just to be clear, @anangl is the ADC maintainer and has the last word about this for what I'm concerned)

anangl commented on drivers/adc/adc_context.h:
```diff
@@ -165,6 +174,10 @@ static inline int adc_context_wait_for_completion(struct adc_context *ctx)
 
 static inline void adc_context_complete(struct adc_context *ctx, int status)
 {
+#ifdef ADC_CONTEXT_ENABLE_ON_COMPLETE
+	adc_context_on_complete(ctx, status);
```
May I ask why is it needed to add this here? It's the driver who calls `adc_context_complete()`, so if it needs to perform some additional operations at that point why cannot it do this directly? What's the benefit of doing this through the adc_context layer?

heinwessels replied:
This is because `adc_stm32_teardown_channels` needs to happen after all sampling is finished. Doing it in `adc_context_on_sampling_done` is not enough, because that can trigger another `adc_context_start_sampling`, which will then sample the ADC _after_ the channel are tore down, without reconfiguring it again.

You can find the conversation that sparked this change [here](https://github.com/zephyrproject-rtos/zephyr/pull/53534#discussion_r1063814975).

anangl replied:
Okay, I understand. Thanks.

heinwessels commented:
> Yeah but the sample is generic, it lives in `samples/drivers/adc`. Not to say it has to support every option, but it makes sense to have samples for bot modes.

Ah, now I understand :) 

If @anangl wants some extra protection in the sample app then I can add it, but IMO it's up to the driver to validate it's input. If the driver receives multiple channels and it can't support it then it should `return -ENOTSUP; ` as mentioned, which the sample app will still gracefully handle.

anangl commented on samples/drivers/adc/src/main.c:
```diff
@@ -51,29 +58,35 @@ void main(void)
+		channels |= BIT(adc_channels[i].channel_id);
 	}
 
+	/* The sequence settings will only depend on the first channel.
+	 * Any settings on other channels will be ignored.
+	 */
+	(void)adc_sequence_init_dt(&adc_channels[0], &sequence);
+
+	sequence.channels = channels;
```
This would essentially change the behavior of the sample and it would actually break the sample for the `nrf52840dk_nrf52840` board which presents usage of [different resolution and oversampling settings for particular channels](https://github.com/zephyrproject-rtos/zephyr/blob/10130d8e6de8c3889f42d498d3de6115db8f2d35/samples/drivers/adc/boards/nrf52840dk_nrf52840.overlay#L18-L35), what is also [mentioned in the sample description](https://github.com/zephyrproject-rtos/zephyr/blob/10130d8e6de8c3889f42d498d3de6115db8f2d35/samples/drivers/adc/README.rst#L29-L34).

I think it's better to add a separate sample if you want to showcase multichannel sequences.

heinwessels replied:
I don't mind simply not supplying a sample. I only added the changes because I used it during development. So would you mind if we simply drop any changes to `samples/`?

anangl replied:
Sure, let's just drop those changes.

erwango replied:
@heinwessels What about adding a new multiple channel sample ?

heinwessels replied:
@erwango if a sample is not required then I'd rather continue on my other Zephyr PRs that's still broken ;). There's also not that big of a difference between sampling a single channel and multiple channels.

heinwessels replied:
Changes to `samples/drivers/adc/` removed by dropping that commit.

fabiobaltieri replied:
It was good value though, it's just that it would not run on all boards so it should not be the only sample provided. I think it'd be cool if you feel like bringing it back as a separate test in a followup PR since you had the work done already, just modified with a filter and config that is known to work that people can add others.

xyzzy42 replied:
I've had to explain how multi-channel sampling works more than once.  +1 to bringing the example back in another PR.
